address-space-controller/src/main/java/io/enmasse/controller/ControllerChain.java

```diff
@@ -87,59 +87,65 @@ public void onUpdate(List<AddressSpace> resources) throws Exception {
+                AddressSpace addressSpace = new AddressSpaceBuilder(original).build();
+                final String addressSpaceName = addressSpace.getMetadata().getNamespace() + ":" + addressSpace.getMetadata().getName();
 
-                log.info("Checking address space {}", addressSpaceName);
-                for (Controller controller : chain) {
-                    log.info("Controller {}", controller);
-                    log.debug("Address space input: {}", addressSpace);
-                    Controller.ReconcileResult result = controller.reconcileAnyState(addressSpace);
-                    addressSpace = result.getAddressSpace();
+                try {
 
-                    // If instructed to requeue, break loop and let the comparison of original vs current
-                    // determine if we need to persist anything. Once persisted, it is assumed that another
-                    // reconciliation event will occur in the near future to allow the reconciliation of this
-                    // AddressSpace to continue.
-                    if (result.isPersistAndRequeue()) {
-                        break;
+                    log.debug("Controller chain input: {}", original);
+
+                    log.info("Checking address space {}", addressSpaceName);
+                    for (Controller controller : chain) {
+                        log.info("Controller {}", controller);
+                        log.debug("Address space input: {}", addressSpace);
+                        Controller.ReconcileResult result = controller.reconcileAnyState(addressSpace);
+                        addressSpace = result.getAddressSpace();
+
+                        // If instructed to requeue, break loop and let the comparison of original vs current
+                        // determine if we need to persist anything. Once persisted, it is assumed that another
+                        // reconciliation event will occur in the near future to allow the reconciliation of this
+                        // AddressSpace to continue.
+                        if (result.isPersistAndRequeue()) {
+                            requeue = true;
+                            break;
+                        }
                     }
-                }
 
-                log.debug("Controller chain output: {}", addressSpace);
+                    log.debug("Controller chain output: {}", addressSpace);
 
-                if (hasAddressSpaceChanged(original, addressSpace)) {
-                    log.debug("Change detected. Executing update.");
-                    if (!this.addressSpaceApi.replaceAddressSpace(addressSpace)) {
-                        log.info("Unable to persist address space state: {}", addressSpaceName);
+                    if (hasAddressSpaceChanged(original, addressSpace)) {
+                        log.debug("Change detected. Executing update.");
+                        if (!this.addressSpaceApi.replaceAddressSpace(addressSpace)) {
```

Comment:
I would assume the replace call will change the resource version. As the replace call uses `lockResourceVersion`, this will work the first time, but not a second time, without reloading the resource.

Reply:
I didn't test if this change has an effect, but I think you're right. I think this change will anyway not work then, back to the drawing table...

Reply:
I think I have a solution that would work now. If requeue is requested, re-list address spaces, which would get the latest resourceVersion of the address spaces. Given that requeues are infrequent, I think the extra 'hit' on the api server is ok.

Reply:
:+1:

I only think that the logic should be described in some source code comment. Also I think the current comment block (about the requeue) is no longer 100% correct.